### ISSUE TITLE
Training blender_3B model: optimization questions (compress model)

### ISSUE THREAD

**Issue by GraphGrailAi** [facebookresearch/ParlAI#3286]
I have 2 questions regarding blender_3B fine-tuning:
@stephenroller @klshuster 

1) How do i compress model to minimize it's size, now it is 16Gb model file, but original is 5Gb. (trained with `--optimizer adam adafactor` and `--dict-tokenizer bpe` ). Why - to lower GPU memory and cost

2) I was not able to fine-tune 3B model with `--dict-tokenizer bytelevelbpe` and adam optimizer (recipe from https://parl.ai/projects/recipes/ with change -t fromfile:parlaiformat --fromfile_datapath ./all.txt) - an error occured.
so the working training string is

`python parlai/scripts/train_model.py -t fromfile:parlaiformat --fromfile_datapath ./all.txt --multitask-weights 1,3,3,3 -veps 0.25 --attention-dropout 0.0 --batchsize 8 --model transformer/generator --embedding-size 2560 --ffn-size 10240 --variant prelayernorm --n-heads 32 --n-positions 128 --n-encoder-layers 2 --n-decoder-layers 24 --history-add-global-end-token end --delimiter '  ' --dict-tokenizer bpe  --dropout 0.1 --fp16 True --init-model zoo:blender/reddit_3B/model --dict-file zoo:blender/reddit_3B/model.dict --label-truncate 128 --log_every_n_secs 10 -lr 7e-06 --lr-scheduler reduceonplateau --lr-scheduler-patience 3 --optimizer adafactor --relu-dropout 0.0 --activation gelu --model-parallel true --save-after-valid True --text-truncate 128 --truncate 128 --warmup_updates 100 --fp16-impl mem_efficient --update-freq 2 --gradient-clip 0.1 --skip-generation True -vp 10 -vmt ppl -vmm min --model-file /tmp/test_train_27B`

What is the difference in terms of talking quality between adam and adafactor (or maybe we need other)?
I got 33 ppl, but adam optimizer need 2(or more) GPU tesla v100 32gb

**Comment by stephenroller** [comment]
1. Model files contain optimizer state, in case we want to resume training. You may use `parlai vacuum -mf model_file` in order to strip extraneous information from the model file. That should significantly reduce file size and loading time
2. Your command isn't right. `--dict-tokenizer bpe` is definitely wrong for reddit_3B.

I don't think we compared Adam/Adafactor for fine-tuning, only pre-training. It made a big difference in pre-training. Two equal models got an improvement of like 14ppl -> 13ppl, which is pretty big improvement in this region.

33 PPL is quite awful. We get ~6-10 on various conversational datasets. I'm pretty sure your issue is that your tokenizer is wrong.

**Comment by klshuster** [comment]
I am not sure if this is just github autocorrect but i would also point out that you have your `--delimiter` set to a single space; it should be a double space

**Comment by stephenroller** [comment]
Lol the delimiter is... tricky.  Single space is actually more correct according to @namangoyal's findings, but it matters not.

**Comment by GraphGrailAi** [comment]
I have another model with ppl=19, and strange thing, it's quality nearly the same as ppl=33, not sure is it depends on dataset volume or training time.

I do implemented custom message_history that manage dialogue flow for user_id, what i faced is that: if i use interactive.py to talk with model it terminal - all fine. But, if i use message_history method i see "ignore effect" - sometimes messages are ignored by model, mostly 3 times and model answer like it do not receive message from me(continue from previous message), but they are not disappeared - if i post the same message 3 times model start to react on them, but suddenly can ignore further.
The strange is that: if i use original 3B model no such behaviour occurred.
